source/common/upstream/ring_hash_lb.cc

```diff
@@ -78,16 +82,47 @@ RingHashLoadBalancer::Ring::Ring(
   const uint64_t min_ring_size =
       config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, 1024) : 1024;
 
+  // Compute the "effective" weight of each host as the product of its own weight and the locality
+  // weight, if given. Sum these effective weights and find their greatest common denominator for
+  // normalization, so that we can size the ring appropriately.
+  // NOTE: GCD is associative, same as std::max() for example. To compute GCD(a, b, c, ...), we
+  //       GCD(GCD(GCD(GCD(0, a), b), c), ...).
+  uint64_t sum = 0;
+  uint32_t gcd = 0;
+  std::unordered_map<HostConstSharedPtr, uint32_t> effective_weights;
+  for (uint64_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      auto host_weight = host->weight();
+      ASSERT(host_weight != 0);
+      // TODO(mergeconflict): C++17 introduces std::shared_ptr::operator[], so we don't have to
+      //                      dereference locality_weights explicitly.
+      auto locality_weight = locality_weights == nullptr ? 1 : (*locality_weights)[i];
+      ASSERT(locality_weight != 0);
+
+      auto effective_weight = host_weight * locality_weight;
+      sum += effective_weight;
+      // TODO(mergeconflict): C++17 introduces std::gcd in <numeric>.
+      gcd = std::__gcd(gcd, effective_weight);
+      effective_weights[host] = effective_weight;
+    }
+  }
+
+  // We can't do anything sensible with no hosts.
+  if (sum == 0) {
+    return;
+  }
+  sum /= gcd;
+
   uint64_t hashes_per_host = 1;
-  if (hosts.size() < min_ring_size) {
-    hashes_per_host = min_ring_size / hosts.size();
-    if ((min_ring_size % hosts.size()) != 0) {
+  if (sum < min_ring_size) {
+    hashes_per_host = min_ring_size / sum;
+    if (min_ring_size % sum != 0) {
       hashes_per_host++;
     }
   }
 
   ENVOY_LOG(info, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size, hashes_per_host);
-  ring_.reserve(hosts.size() * hashes_per_host);
+  ring_.reserve(sum * hashes_per_host);
```

Comment:
Overall, looks great, but one question.

In this change, when we have a ring size of 16K and 12K hosts with a range of hosts, we could end up having the ring size be some large number, e.g. multiple MB. This will also fluctuate as the management server, unaware of the implications on memory use, makes small adjustments to weighting, causing things that used to divide cleanly to no longer.

Would it be useful to add a maximum bound on ring size and sacrifice granularity when hit? @mattklein123 for thoughts.

Reply:
I haven't looked at the PR in detail, but IMO the ring size should be fixed, as specified in config, and we should size the replication of the hosts within that size.

Reply:
Looking at https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cds.proto#envoy-api-msg-cluster-ringhashlbconfig, we express this as a minimum today, and then say there is a fixed 8MB maximum. How about we do:
* `minimum_ring_size` with existing semantics as today (the ring will be at least this big and contain every host at least once).
* `maximum_ring_size` defaults to 8MB but might be set lower. E.g. we could have 16KB for both min and max in some configurations.
?

Reply:
I think the documentation as written might be a bit too easy to misinterpret... It says "This field is limited to 8M to bound resource use," meaning that there's an upper bound on `minimum_ring_size`, but there isn't currently any upper bound on the actual ring size. So I'd definitely be in favor of adding `maximum_ring_size`, imposing an actual upper bound on resource use.

Note that Ketama is probabilistic — the more hashes per host, the more closely its partitioning of the key space will resemble the desired weights (even if all the weights are 1) — so we should be pretty explicit in the documentation about configuring ring size bounds appropriately for the cluster.

Reply:
Yes, please add a maximum_ring_size and provide a sensible default. Previously, it was very easy to intuit memory usage, now it isn't.

Reply:
This is implemented and documented in the latest change, let me know what you think.